### DISCUSSION TITLE
Add (and improve comparing to stable) bookmarks to the Editor

### DISCUSSION THREAD

**Discussion by Misty-Owl** [ppy/osu#29169] [Ideas]
The ability to add markers (bookmarks) on Editor timeline, for which you can set a color, title and description. Just like in any good timeline based editing software.

The rough example of how it could look like is shown in the picture below:

![osu marker example](https://github.com/user-attachments/assets/9efebdc6-dcfe-4b7d-b398-8b75e3e22a34)

**Comment by AncuL001** [comment]
Additional idea: body background on the marker hint could have the same color tone as the marker

**Comment by asmorodinov** [comment]
Hi, I like this suggestion, and it seems like a lot of other people also like it, so I want to try to bump this discussion.

## More information about use cases

1. In general, the idea is to have a place within the editor to add and view notes that are connected to certain timestamps. Bookmarks seem like an ok place to do that, but maybe another system for that would be better (especially if it gets more complicated than just colors and text).

2. Mappers can write notes themselves, or they can get notes from other mappers (e.g. map feedback) and then import them into their map (second feature is more complicated to implement, but still doable). For the second feature we would need to store bookmark author for each bookmark (to distinguish them). Then we could implement importing bookmarks from other mappers, merging them into the current file, selecting based on the author, etc.

3. Colors by themselves are also nice, they allow for example to mark calm parts of a song with green bookmarks, and intense parts with red, or whatever else the mapper can think of. In case of multiple people mapping together, they can mark their respective sections of the map with bookmarks with custom colors that they choose.

4. There are some additional ideas, like loading comments from beatmap discussion tab and importing them as bookmarks with text. But this is a more complicated feature, and requires online implementation.

For now, I think we can add: Color, Title and Description to bookmarks (and maybe Author as well). This would be the simplest approach, and more complicated features can be added later if needed (e.g. importing other mappers bookmarks into your own map, but this can also be made by a third-party script or manually).

## Implementation details

Here is a little bit of my initial thoughts on implementing this (mostly from the backend implementation standpoint for now, not the UI).

1. Right now, bookmarks are stored as comma-separated list of integers (In the Editor->"Bookmarks" section of .osu file), and loaded to / saved from `BindableList<int>` at runtime.

2. Each addition and removal of a bookmark requires a binary search + Insert/Remove + OrderBy (sort). O(n log n) operations in total, potentially slow, but probably good in enough in practice.

3. I suggest to replace `BindableList<int>` with `BindableList<BookmarkData>`, or `BindableMap<int, BookmarkData>` (for faster insert, and to not have to sort the container every time). We can then store bookmark color, title and description inside `BookmarkData` (and potentially author name). 

4. In the osu file, this info would be stored inside an additional field, e.g. Editor -> "BookmarksData" (it would be separate from "Bookmarks" for backward-compatibility reasons). We need to decide, how to serialize `BookmarkData` itself, since it contains user-input. Maybe we can use a separator and forbid to use it in the text of bookmark, or use some other method (like escaping certain characters). osu!stable and older lazer clients will simply ignore this "BookmarksData" field.

## Notes

Please let me know, what do you think about this idea. I can implement this and make a PR, if it is ok

**Reply by bdach** [reply]
None of this is probably happening until a new beatmap format is introduced.

**Reply by asmorodinov** [reply]
Is there anything I can help with, for now? There is this discussion about new format, but last message was in 2023 - https://github.com/ppy/osu/discussions/12976 (I understand that this is probably low priority at the moment).

Is it ok if I just add a new field in the current format?

I would like to implement this, at least for personal use, but ideally for other people too.

**Reply by peppy** [reply]
I'd be fine with extensions to bookmarks being added to the `.osu` format FWIW. It should be pretty simple to do it in a way that doesn't annoy stable.

**Reply by cihe13375** [reply]
I am a bit worried that some people may insert harmful contents into bookmarks, and BNs/admins have to check with lazer to find them out. Also now modders need to check mappers' comments in lazer before posting mods, to see if their points are already considered by the mapper. I think the better route should be loading comments from (and upload comments into) beatmap discussion, instead of injecting this information into the map itself.

**Reply by asmorodinov** [reply]
Thank you both for your responses, and bringing up good questions.

You are right, there is already a "Notes" category in the beatmap discussion tab, maybe mappers can use it for their personal comments about the map. Or a new category can be added as well.

However, I think that comments within a locally saved beatmap file also have their use cases:
1. Before the map is posted to the website, users still should be able to add comments. They cannot do that with a discussion tab (there is no page for the map on the website yet).
2. You should be able to map without internet access.
3. While creating a map, mappers constantly make changes, and they don't want to make these intermediate changes public (including the comments). They only want to make the final map public by explicitly uploading it to the website.
4. You have to store local comments somewhere, .osu file (or a separate file within .osz) seems like a reasonable place.

So, I think it is logical to distinguish local comments (in bookmarks, or in a separate list) and online comments (in the discussion tab).
Both of those types of comments can be viewed or edited in the same tab in the editor, but it is useful to distinguish them for understanding the pipeline better.

To avoid issues with moderation and modding, I think the process might look something like this:
1. User is making a map, they can add/delete/modify local comments however they want. These comments are exported and saved locally, but not uploaded to the server.
2. User decides that the map is good enough and wants to upload it. Then they can choose between the three options:
    1. Post all of their local comments in the beatmap discussion tab (in the Notes category, for example).
    2. If they are not comfortable with publicly sharing their personal comments (or think that they may be considered offensive to someone), then they can choose to not post their comments in the discussion tab.
   3. User can select some of the comments to post in the discussion tab (middle ground between posting everything and posting nothing, most flexible option for the user).

Regardless of what the user chooses, local comments within a local osu file are not uploaded to the server. Remote file will not have these comments, but local file will. There might be a potential issue with synchronization because of this (lazer would think that the local file differs from the online file and would prompt to redownload it), but it probably can be solved somehow.
This solves the moderation issue and modders issue.

A (maybe better) alternative would be to include comments in the server file, but also show them in the discussion tab. User still might choose not to upload some of the local comments (they will not be added to the online file and to the discussion tab). This is a bit more tricky, because you can not delete online comments, but you can delete local comments, so those lists might diverge.

3. For the online comments, I think there should be a separate tab within the editor where they can be viewed and replied to. These will only be available once the map is uploaded to the website.

Note: third part is more complicated to implement, maybe we should start with just the first two parts. And always force option 2.ii for now. If users want to share their local comments they would have to export the file and share the file (not ideal, but maybe it is ok for now).

Another note: implementing only colors for bookmarks is even simpler, does not require moderation from BNs and is still a useful feature by itself. For example, users can mark sections with different intensities with one color, specific sounds with different colors.
Maybe start with just implementing colors?

Please let me know what do you think.

**Reply by bloeiendelgtbbt** [reply]
pleeaseee do that!!! it would be muuuch handy to map!

**Comment by bloeiendelgtbbt** [comment]
YEEESSSSS PLEASEEEEEEE UP UP UPPPPP
